rotkehlchen/db/evmtx.py

```diff
@@ -154,6 +155,67 @@ def get_evm_internal_transactions(
 
         return transactions
 
+    def assert_tx_data_is_pulled(
+            self,
+            cursor: 'DBCursor',
+            evm_inquirer: 'EvmNodeInquirer',
+            tx_hash: 'EVMTxHash',
+            relevant_address: Optional['ChecksumEvmAddress'],
+    ) -> None:
+        """Makes sure that the receipt of the given transaction is in the database.
+        If not, pulls it and stores it in the database.
+
+        May raise:
+        - RemoteError if there is a problem querying the data source.
+        """
+        tx_receipt_data = cursor.execute(
+            'SELECT receipts.tx_id FROM evm_transactions AS transactions '
+            'INNER JOIN evmtx_receipts AS receipts ON transactions.identifier = receipts.tx_id '
+            'WHERE transactions.tx_hash = ?',
+            (tx_hash,),
+        ).fetchone()
+
+        if tx_receipt_data is not None:
+            return  # all good, tx receipt is in the database
+        
+        transaction, raw_receipt_data = evm_inquirer.get_transaction_by_hash(tx_hash)
+        # with self.db.conn.write_ctx() as write_cursor:
+        self.add_evm_transactions(
+            write_cursor=cursor,
+            evm_transactions=[transaction],
+            relevant_address=relevant_address,
+        )
+        self.add_receipt_data(
+            write_cursor=cursor,
+            chain_id=evm_inquirer.chain_id,
+            data=raw_receipt_data,
+            )
+
+    def get_or_create_transaction(
```

Comment:
Okay so I left you a comment about the cursor name being changed to write_cursor since you pass a writing one but that is a problem.

You should not pass a write_cursor here as passing a write_cursor forces you to have one open for long. If you recall from the chat we had that is a bad thing. Haning an open Database transaction for long can lead to errors and deadlocks.

What you should do instead is:
- Pass a read cursor, and use it to read from the DB
- If needed pull the transaction/receipt data from the remote (this is what may take long - as it's a remote call). You absolutely **do not** want an open write cursor, thus an open DB transaction while a remote query is happening
- Once done and you have the data, then and only then open a write cursor, which opens a DB transaction to write it.

If you need to do multiple calls, then gather the data in memory after each call and then write them all together under a single write_cursor once you are done.